### PR TITLE
ci: Add workflows to push, pull, and validate docker images

### LINKED ISSUE #1266: Workflow to publish container to public registy
Registry needs to publically resolvable - it can be Docker's, [Github](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry)'s, Web3 one, etc.

### LINKED ISSUE #1674: Docker container CI
We currently do not publish Docker containers to any public registries.  We should add an automated build step for publishing the latest docker images to both GitHub Packages & DockerHub.

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #1674 
Resolves #1266

## Description

This PR adds a GitHub Action workflow that builds and publishes DefraDB container images to DockerHub & GitHub container registries.

Push Docker images to registries Job:
https://github.com/sourcenetwork/defradb/actions/runs/5651002188

Pull Docker image Job:
https://github.com/sourcenetwork/defradb/actions/runs/5651962572

Validate containerfile job:
https://github.com/sourcenetwork/defradb/actions/runs/5652056549

GitHub Container Image:
https://github.com/sourcenetwork/defradb/pkgs/container/defradb

DockerHub Container Image:
https://hub.docker.com/repository/docker/sourcenetwork/defradb/general

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Manual

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

**Review by fredcarle — COMMENTED** [review]
It seems good to me but I'll let Shahzad approve it as he's much better at CI stuff than I am.

**Comment by fredcarle** [comment]
Not sure if you saw this but this draft PR was trying to do something similar #1350.
The associated issue is #1266.

**Comment by nasdf** [comment]
> Not sure if you saw this but this draft PR was trying to do something similar #1350. The associated issue is #1266.

My bad, I should have searched for that before starting. I'll reference that open issue here as well.

**Comment by shahzadlone** [comment]
Hey Keenen let me know when this is unblocked as I still see PR description says blocked until secrets are added (are the secrets not working?)

Please link the last execution run of this actions successfull execution in the PR.

Would be nice to also see the link to the registery when the build gets pushed to.

**Comment by nasdf** [comment]
> Hey Keenen let me know when this is unblocked as I still see PR description says blocked until secrets are added (are the secrets not working?)
> 
> Please link the last execution run of this actions successfull execution in the PR.
> 
> Would be nice to also see the link to the registery when the build gets pushed to.

My bad, I forgot to remove that after you added the secrets.  I'll attempt a test run of it now.

File: .github/workflows/build-then-deploy-docker.yml
Comment (shahzadlone) [inline .github/workflows/build-then-deploy-docker.yml:16]: Issue: My hunch is that because you are using the fork-flow to test the trigger, `pull_request` has no access to secrets.

**Review by shahzadlone — CHANGES_REQUESTED** [review]

**Comment by shahzadlone** [comment]
suggestion: You could try testing two ways:
1) try `pull_request_target`
2) try opening a dummy draft pr with branch flow instead of forkflow

Note: secrets will be available with tag push

File: .github/workflows/build-then-deploy-docker.yml
Comment (shahzadlone) [inline .github/workflows/build-then-deploy-docker.yml:21]: question: Do we actually want a new image pushed to docker for every commit on develop? and on master?

todo: IMO pushing only for every tag might be sufficient, or if we do want docker container for the latest commit we can have a different action with that capability to build at the head commit on develop when we trigger that action manually.

```suggestion
```

Comment (nasdf) [reply .github/workflows/build-then-deploy-docker.yml]: That's a good point. If we need dev images we can also manually build. I applied your change.

File: .github/workflows/build-then-deploy-docker.yml
Comment (shahzadlone) [inline .github/workflows/build-then-deploy-docker.yml:13]: todo: `Build` signals to me this workflow cares or tests some build step.
Would like a name more descriptive of the core purpose of this workflow.
 
```suggestion
name: Push Docker Image To Registries Workflow
```

File: .github/workflows/build-then-deploy-docker.yml
Comment (shahzadlone) [inline .github/workflows/build-then-deploy-docker.yml:1]: todo: Following the other suggestion would name the file to: `push-docker-image-to-registries.yml`

File: .github/workflows/build-then-deploy-docker.yml
Comment (shahzadlone) [inline .github/workflows/build-then-deploy-docker.yml:22]: todo: We usually put `job` in the ending of every job's name (helps distinguish steps, workflows, and jobs in various places).

```suggestion
    name: Push docker image to registries job
```

File: .github/workflows/build-then-deploy-docker.yml
Comment (shahzadlone) [inline .github/workflows/build-then-deploy-docker.yml:56]: question: Does it notify with a fail if a build was to fail?

File: .github/workflows/build-then-deploy-docker.yml
Comment (shahzadlone) [inline .github/workflows/build-then-deploy-docker.yml:55]: nitpick:

```suggestion
      - name: Push docker images as build succeeds
```

**Review by shahzadlone — CHANGES_REQUESTED** [review]

File: .github/workflows/build-then-deploy-docker.yml
Comment (shahzadlone) [inline .github/workflows/build-then-deploy-docker.yml:59]: question: What was the difference between `file` and `path`.

Comment (nasdf) [reply .github/workflows/build-then-deploy-docker.yml:59]: I made a typo when first creating it. The `path` param doesn't actually exist.  The `file` param is equivalent to `docker build -f tools/defradb.containerfile`

Comment (shahzadlone) [reply .github/workflows/build-then-deploy-docker.yml:59]: Makes sense, that's what I thought! Thanks

File: .github/workflows/build-then-deploy-docker.yml
Comment (nasdf) [reply .github/workflows/build-then-deploy-docker.yml:56]: Yes, I believe so https://github.com/docker/build-push-action/blob/master/src/main.ts#L54

File: .github/workflows/push-docker-image-to-registries.yml
Comment (shahzadlone) [inline .github/workflows/push-docker-image-to-registries.yml:21]: nitpick:

```suggestion
  push-docker-image-to-registries:
```

**Review by shahzadlone — CHANGES_REQUESTED** [review]
Some thoughts on the new approach:

We want 2 things on top of the previous basic implementation which did the basic build and push:

1) We wanted to make sure our container file is always in a valid state (it is possible to be in a valid state but pushes something that is missing deps, as we had with the shared lib recently).

2) We want to test that we can pull the pushed docker image (from either one of the registry should suffice) and attempt to run the binary, to ensure there are no error like we had with the shared lib previously.

I see both 1) and 2) as separate features from each other and from what we had previously.

Problem with the current approach:
- It combines the pushing of the image and validating the container file (1) in one step, meaning we will only be able to validate the container file on every tag-push (IMO it needs to be in valid state always). Here is an idea for a separate workflow that will run every time the container file is touched.

```yaml
// defradb/.github/workflows/validate-containerfile.yml
# Copyright 2023 Democratized Data Foundation
# INSERT REST OF HEADER

name: Validate Containerfile Workflow

on:
  pull_request:
    branches:
      - master
      - develop
    paths:
      - '.github/workflows/validate-containerfile.yml'
      - 'tools/defradb.containerfile'

  push:
    branches:
      - master
      - develop

jobs:
  validate-containerfile:
    name: Validate containerfile job

    runs-on: ubuntu-latest

    steps: # Insert the steps to validate the containerfile
```

Note: You can also keep the current new validation step you added to additionally check before pushing as well (as it still will only build one for each platform and reuse the build while pushing).

As for (2) you can have an action spin of after this action is completed automatically if you do something like this:
```yaml
# File:  defradb/.github/workflows/pull-docker-image-workflow.yml

# Copyright 2023 Democratized Data Foundation
# INSERT REST OF HEADER

name: Pull Docker Image Workflow

on:
  workflow_run:
    # Warning: this workflow must NOT:
    # - interact with any new code.
    # - checkout new code.
    # - build/compile anything (only pull).
    # - make any indirect calls (i.e. make xyz, or npm install, etc.)
    # Note this workflow:
    # - will use the base's (or default) workflow file's state.
    # - doesn't run on the PR or the branch coming in, it runs on the default branch.
    # - has read-write repo token
    # - has access to secrets
    workflows: ["Push Docker Image To Registries Workflow"]
    types:
      - completed

jobs:
  pull-docker-image:
    name: Pull docker image job

    if: ${{ github.event.workflow_run.conclusion == 'success' }}

    runs-on: ubuntu-latest

    steps: #INSERT STEPS 
```

File: .github/workflows/push-docker-image-to-registries.yml
Comment (nasdf) [reply .github/workflows/push-docker-image-to-registries.yml]: Added platforms.

File: .github/workflows/push-docker-image-to-registries.yml
Comment (nasdf) [reply .github/workflows/push-docker-image-to-registries.yml:67]: Probably a good idea to build for both. Added.

**Comment by nasdf** [comment]
> Note: You can also keep the current new validation step you added to additionally check before pushing as well (as it still will only build once for each platform and reuse the build while pushing).

I attempted to keep this functionality but I ran into an issue with Docker not allowing export of multi platform images.

**Comment by shahzadlone** [comment]
> > Note: You can also keep the current new validation step you added to additionally check before pushing as well (as it still will only build once for each platform and reuse the build while pushing).
> 
> I attempted to keep this functionality but I ran into an issue with Docker not allowing export of multi platform images.

Lets just stick to default platform for now. We can look into adding more platforms later then.

Lmk if you need any clarification on the other 2 workflows i mentioned.

File: tools/defradb.containerfile
Comment (shahzadlone) [inline tools/defradb.containerfile:10]: thought: Maybe we can get by not building lens deps here as they might be only dev dependencies. I will make a ticket and it can be taken care of outside the scope of this PR.
New Issue here:
https://github.com/sourcenetwork/defradb/issues/1707

**Review by shahzadlone — APPROVED** [review]
LGTM! This is some awesome work, appreciate you adding all the changes just shared one `thought` and still wondering if you are doing the one `nitpick` I raised before (but is non-blocking)

File: .github/workflows/push-docker-image-to-registries.yml
Comment (shahzadlone) [reply .github/workflows/push-docker-image-to-registries.yml:24]: @nasdf Missed ?

**Comment by shahzadlone** [comment]
nitpick: Can update title to `ci: Add workflows to push, pull & validate docker`

File: .github/workflows/push-docker-image-to-registries.yml
Comment (nasdf) [reply .github/workflows/push-docker-image-to-registries.yml]: Sorry, totally missed this one. Fixed.

**Pull request merged**
